src/axolotl/utils/callbacks.py

```diff
@@ -317,3 +320,149 @@ def on_evaluate(
+                        batch_pos_ids = batch["position_ids"].tolist()
+                    else:
+                        batch_pos_ids = [None] * len(batch["input_ids"])
+
+                    prompt_token_ids_list = []
+                    completion_token_ids_list = []
+
+                    for input_ids_all, labels_all, pos_ids in zip(
+                        batch_input_ids, batch_labels, batch_pos_ids
+                    ):
+                        if pos_ids is None:
+                            pos_ranges = [(0, len(input_ids_all) - 1)]
+                        else:
+                            pos_ranges = find_ranges(pos_ids)
+
+                        for pos_range in pos_ranges:
+                            start, end = pos_range
+                            if start == end:
+                                continue
+
+                            input_ids = input_ids_all[start : end + 1]
+                            labels = labels_all[start : end + 1]
+
+                            tokens_without_loss = labels == IGNORE_INDEX
+                            tokens_with_loss = labels != IGNORE_INDEX
+                            tokens_exclude_padding = input_ids != tokenizer.pad_token_id
+
+                            prompt_token_includes = (
+                                tokens_without_loss & tokens_exclude_padding
+                            )
+
+                            prompt_token_ids = input_ids[prompt_token_includes]
+                            prompt_token_ids_list.append(prompt_token_ids)
+
+                            completion_token_ids = input_ids[tokens_with_loss]
+                            completion_token_ids_list.append(completion_token_ids)
+
+                    prompt_texts = tokenizer.batch_decode(
+                        prompt_token_ids_list, skip_special_tokens=True
+                    )
+                    completion_texts = tokenizer.batch_decode(
+                        completion_token_ids_list, skip_special_tokens=True
+                    )
+
+                    with torch.no_grad():
+                        generation_config = GenerationConfig(
+                            max_new_tokens=max_new_tokens,
+                            bos_token_id=tokenizer.bos_token_id,
+                            eos_token_id=tokenizer.eos_token_id,
+                            pad_token_id=tokenizer.pad_token_id,
+                            do_sample=False,
+                            use_cache=True,
+                            return_dict_in_generate=True,
+                            output_attentions=False,
+                            output_hidden_states=False,
+                            output_scores=False,
+                        )
+
+                        prompt_encoding = tokenizer(
+                            prompt_texts, padding=True, return_tensors="pt"
+                        ).to(self.cfg.device)
+                        predictions = trainer.model.generate(
```

Comment:
`trainer.prediction_step(...)` might be easier to use

Reply:
It was. However, I had that previously and it appeared to output strange predictions. At one point I actually had both to compare and only `model.generate` was useful.

I'm struggling to find the WandB report. I'll try again and see how it goes.

Reply:
Top is `trainer.prediction_step` and bottom is `trainer.model.generate`:

![image](https://github.com/OpenAccess-AI-Collective/axolotl/assets/1885333/380db98a-26cd-410d-8229-67e6074d6d49)

Not sure if this is correct though? I originally showed both, I could add it too?

Reply:
Added both now

Reply:
<img width="1974" alt="image" src="https://github.com/OpenAccess-AI-Collective/axolotl/assets/1885333/f2475c49-1915-4831-b549-bffd637258f2">